corehq/apps/hqadmin/management/commands/dump_domain_data.py

```diff
@@ -0,0 +1,142 @@
+from collections import OrderedDict
+
+from django.apps import apps
+from django.conf import settings
+from django.core import serializers
+from django.core.management.base import BaseCommand, CommandError
+from django.db import router
+
+from corehq.sql_db.config import partition_config
+
+app_labels = {
+    'locations.LocationType': 'domain',
+    'locations.SQLLocation': 'domain',
+    'form_processor.XFormInstanceSQL': 'domain',
+    'form_processor.XFormAttachmentSQL': 'form__domain',
+    'form_processor.XFormOperationSQL': 'form__domain',
+}
+
+
+class Command(BaseCommand):
+    help = "Dump a domain's data to disk."
+    args = '<domain>'
+
+    def add_arguments(self, parser):
+        parser.add_argument('-e', '--exclude', dest='exclude', action='append', default=[],
+            help='An app_label or app_label.ModelName to exclude '
+                 '(use multiple --exclude to exclude multiple apps/models).')
+        parser.add_argument('--indent', default=None, dest='indent', type=int,
+                            help='Specifies the indent level to use when pretty-printing output.')
+        parser.add_argument('-o', '--output', default=None, dest='output',
+            help='Specifies file to which the output is written.')
+
+    def handle(self, domain, **options):
+        format = 'json'
+        indent = options.get('indent')
+        excludes = options.get('exclude')
+        output = options.get('output')
+        show_traceback = options.get('traceback')
+
+        _check_serialization_format(format)
+
+        excluded_apps, excluded_models = _get_excluded_apps_and_models(excludes)
+        app_list = _get_app_list(excluded_apps)
+
+        def get_objects():
+            # Collate the objects to be serialized.
+            for model in serializers.sort_dependencies(app_list.items()):
+                if model in excluded_models:
+                    continue
+
+                using = router.db_for_read(model)
+                if settings.USE_PARTITIONED_DATABASE and using == partition_config.get_proxy_db():
+                    using = partition_config.get_form_processing_dbs()
+                else:
+                    using = [using]
+
+                for db_alias in using:
+                    if not model._meta.proxy and router.allow_migrate_model(db_alias, model):
+                        objects = model._default_manager
+
+                        label = '{}.{}'.format(model._meta.app_label, model.__name__)
+                        filter_kwarg = app_labels[label]
+
+                        queryset = objects.using(db_alias) \
+                            .filter(**{filter_kwarg: domain}) \
+                            .order_by(model._meta.pk.name)
+
+                        for obj in queryset.iterator():
+                            yield obj
+
+        try:
+            self.stdout.ending = None
+            stream = open(output, 'w') if output else None
```

Comment:
Will probably have to split this into multiple files and zip it up so that we don't have to read the entire file into memory when loading.